### PR TITLE
Migrating SpaceX API to GraphQL 🚀

### PR DESCRIPTION
**Pull request by swcarlosrj**
Hey there,

I've migrated the actual REST API to GraphQL, as user who wants to know more about SpaceX's public data, I've found out interesting the fact of being able to interactively introspect entrypoints-data & explore full-updated docs, which means that a lot of all non technical users might access to the data too.

You could test the GraphQL API here 👉 https://codesandbox.io/s/l77040m1zm
Also, you could read about the motivations here 👉 https://medium.com/@swcarlosrj/migrating-spacex-api-to-graphql-e1fe69a3a8e7

Loving inspiring work that you guys are doing, hoping that this is useful for y'all & the OSS community!

### CODE REVIEW

**Comment by jhpratt** [comment]
Just reading the blog post. With regard to overfetching, there's actually an optional querystring called `filter`, that will only return the fields you want. Not sure if it's documented or not.

I'm not familiar enough with GraphQL (beyond a concept-level), so I've requested Jake review this one. No idea if he has experience with it, either. I know it's never been discussed on Slack.

**Comment by jakewmeyer** [comment]
Will review when I'm back in town later this weekend

**Comment by jakewmeyer** [comment]
First of all, I appreciate the time and effort you put into this migration, you've been very thorough. However, I do have a few concerns about the potential implications of merging this.

* We already have a GraphQL implementation that someone has implemented [here](https://github.com/emersonlaurentino/spacex-qraphql-api). This PR might be a better fit for that repo.

* After reading the blog post, I'm concerned that this "migration" may lead to an overzealous removal of the REST implementation at some point in the future.

* This PR completely disregards the current eslint style and linting rules.

* Is this going to be maintained regularly?

* This adds additional overhead to deployments, and would require more time to come up with a deployment and caching strategy.

Also want to address some of the points made in the blog post:

* We prevent overfetching by providing a filter querystring that allows users to grab only the data they need.

* Underfetching could be a concern, but multiple requests to commonly cached items is still a shorter request than one unique roundtrip all the way to the database.

* `I just want to explore data & their types without the need of setting anything up` The Docs and Postman collection make it very easy to do this. It lets you explore the responses, and view documentation about optional querystrings. No need to set anything up, it's already available.

* The API is versioned to allow for backwards compatibility, while still providing an upgrade path for users interested in new features. With GraphQL, you bank on the fact that users will pay attention and adjust to deprecated fields.

**TLDR:** Both REST and GraphQL have pros and cons, and it's not as clear cut as you would lead others to believe. I understand that you're supporting the GraphQL community, but this just feels like a marketing stunt. This would've been better as an issue for discussion, instead of wasting time to review a PR that never had the intention of merging.

**Comment by swcarlosrj** [comment]
@jakewmeyer thanks so much for taking your time to answer 😊

Regarding your concerns:
• Thanks for sharing it, I didn't even know that existed.
• Why? They might coexist, not discovering anything new (Github [v3](https://developer.github.com/v3/), Github [v4](https://developer.github.com/v4/))
• It can be fixed, the purpose of this MR is to show a GraphQL implementation to start working with it, there are so many features which would need to be fixed such as type definitions, error handling, testing, etc...
• I'd be 😄to do that!
• Of course as any SW upgrade does but thinking about cost-benefit, would this improve your API?

Regardings points made:
• As far I know you only do for some use cases, is [this](https://api.spacexdata.com/v3/launches?filter=rocket/second_stage/payloads/payload_id,orbit) even possible? I'd really appreciate to have it explained over the docs!
• Underfeching is a real ⚠️here. On the following [example](https://graphqlbin.com/v2/wVplUL) I'd need 4 calls, a parallel async method and a join data function.
• It requires install Postman but anyways, would GraphQL improve your current API UX?
• I'm not an GraphQL continuous evolution expert but I do know that there are several ways to solve this easily, [here](https://speakerdeck.com/xuorig/continuous-evolution-of-graphql-schemas-at-github) a great post about it!

**TLDR**: This is not about REST vs GraphQL, it's about type systems, performance, user experience and such. Hoping that instead of taking this MR as a wasting of time, you've taken it as an opportunity of learning and improving your current API 🤗

As a non representative OSS contributor doesn't make me sense talk about marketing but I'll love to talk about considering people's time & desire to help!

**Comment by jakewmeyer** [comment]
If you're willing to maintain it, I would be willing to merge in the future with some changes 👍 

Maybe coordinate with @emersonlaurentino and see if he has any other suggestions or requests from his repo, he might be able to help maintain as well.

**Comment by swcarlosrj** [comment]
Thanks @jakewmeyer for considering it 😊, I'm already working on type safety and testing. In order to have a 'beta' it shouldn't take longer than a month, let's work together to make it happen. Is there a Slack channel available to ping you guys up?

**Comment by jakewmeyer** [comment]
Closing this, because it will be in a new repo 👍
